### PR TITLE
Port to clap v3

### LINKED ISSUE #298: Replace `gumdrop` with `clap` v3?
There were two primary reasons why Abscissa used `gumdrop` over `clap` v2 originally:

- Previously `clap` had an excessively large number of dependencies, whereas `gumdrop` had no dependencies other than `proc-macro2`, `quote`, and `syn` for its proc macro. It seems `clap-v3` has done a good job of addressing this (see below).
- Custom derive support: `clap`'s previous custom derive crate `structopt` was added as something of an afterthought, had a clunky attribute syntax, and generally felt pretty low quality, whereas `gumdrop` was effectively a "custom derive first" crate and in that regard felt much more cohesive. That said, `clap-v3` now features [`clap_derive`](https://github.com/clap-rs/clap/tree/master/clap_derive) which largely addresses these concerns.

We're now at a crossroads where we can either upgrade to `gumdrop` v0.8 or consider migrating to `clap-v3`. Either approach seems a bit difficult as `gumdrop` v0.8 contains a number of API changes.

The current beta release is [clap-v3 3.0.0-beta.2](https://crates.io/crates/clap/3.0.0-beta.2). They warn against depending on the API until it's stable, however.

Tracking issue for `clap` v3 is here: https://github.com/clap-rs/clap/issues/1037

## Dependencies

Here are the transitive dependencies of `clap-v3` 3.0.0-beta.1 (I haven't fully cross-referenced these with our current ones yet, but I see a lot of commonality):

- ansi_term 0.12.1
- atty 0.2.14
- autocfg 1.0.0
- bitflags 1.2.1
- clap-v3 3.0.0-beta.1
- clap_derive-v3 3.0.0-beta.1
- heck 0.3.1
- hermit-abi 0.1.13
- indexmap 1.4.0
- lazy_static 1.4.0
- libc 0.2.71
- proc-macro-error 0.4.12
- proc-macro-error-attr 0.4.12
- proc-macro2 1.0.18
- quote 1.0.7
- strsim 0.9.3
- syn 1.0.31
- syn-mid 0.5.0
- textwrap 0.11.0
- unicode-segmentation 1.6.0
- unicode-width 0.1.7
- unicode-xid 0.2.0
- vec_map 0.8.2
- version_check 0.9.2
- winapi 0.3.8
- winapi-i686-pc-windows-gnu 0.4.0
- winapi-x86_64-pc-windows-gnu 0.4.0

### PR DESCRIPTION
**Pull request by rsdy**
This is a reasonably straightforward port of Abscissa to Clap v3 using the custom derives.
I removed most of the command subsystem (`EntryPoint`, `Usage`, `Help`) since Clap takes over most of those roles.

To demonstrate how this could work, I also migrated the CLI.

The ergonomics might be improved, as subcommands work differently in Clap than Gumdrop. This is best seen in the [`commands.rs`](https://github.com/iqlusioninc/abscissa/blob/6cbbfe2b1d94f06bce79e11d7e9161147dcea708/cli/src/commands.rs) file of the CLI and [`gen.rs`](https://github.com/iqlusioninc/abscissa/blob/6cbbfe2b1d94f06bce79e11d7e9161147dcea708/cli/src/commands/gen.rs) root command. Embedding subcommands can get tedious quickly.

Another gotcha I found is that `#[derive(Runnable)]` will not work for structs in its current form, so a [manual implementation](https://github.com/iqlusioninc/abscissa/pull/479/files#diff-e2e916547f7e32f6295257919ded6b0976ca9231d2f0554a1a7f070697d0796bR32) is necessary.

If this changeset looks like a viable approach, I can take a look at porting the template.

Resolves #298

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -16,7 +16,7 @@ categories = ["command-line-interface", "config", "rust-patterns"]
 keywords   = ["abscissa", "cli", "application", "framework", "service"]
 
 [dependencies]
-gumdrop = "0.7"
+clap = "3.0.0-beta.1"
```
Comment (tony-iqlusion) [inline cli/Cargo.toml:19]: FYI, `3.0.0-beta.2` is out: https://crates.io/crates/clap/3.0.0-beta.2

**Comment by tony-iqlusion** [comment]
Thanks! This is definitely wanted and we're glad you opened the PR!

Per the `clap` v3 docs:

https://crates.io/crates/clap/3.0.0-beta.2

> We have a 3.0.0-beta.2 prerelease out but we do not give any guarantees that it's API is stable. We do not have a changelog yet which will be written down after we are sure about the API stability. We recommend users to not update to the prerelease version yet and to wait for the official 3.0.

We definitely want to merge a PR like this after the official 3.0 release, but per that message, we don't want to do it beforehand.

**Comment by rsdy** [comment]
@tony-iqlusion Happy to keep ironing out ergonomics issues and templates until Clap stabilises.

**Comment by rsdy** [comment]
I've ported forward all the templates, and local tests pass.

I've tried re-introducing an `EntryPoint` so that the application template doesn't need to deal with `config` and `verbose`, but nether trait objects or generics will work due to the `Clap` derive macro. Overall, having that code generated might slightly increase flexibility, but at the cost of more noise.

**Comment by tony-iqlusion** [comment]
Closing in favor of #562, which builds upon this work.

**Pull request closed without merging**